File: solutions/sliding-window/javascript/0209-minimum-size-subarray-sum.js
/**
 * ### METADATA:
 *
 *
 * ### INTUITION:
 * [This problem requires understanding of sliding window concepts. The key insight is to identify the optimal approach for this specific scenario.]
 *
 * ### APPROACH:
 * 1. **Analyze the problem**: Understand the input constraints and expected output
 * 2. **Choose the right technique**: Apply sliding window methodology
 * 3. **Implement efficiently**: Focus on optimal time and space complexity
 * 4. **Handle edge cases**: Consider boundary conditions and special cases
 *
 * ### WHY THIS WORKS:
 * - This ensures that the solution leverages sliding window principles
 * - This ensures that time complexity is optimized for the given constraints
 * - This ensures that space complexity is minimized where possible
 *
 * ### EXAMPLE WALKTHROUGH:
 * Input:
 * ```
 * target = 7, nums = [2,3,1,2,4,3]
 * ```
 *
 * Step 1: Expand window
 * [2,3,1,2] sum=8 ≥ 7
 * Step 2: Contract
 * [3,1,2] sum=6 < 7
 * Expand: [3,1,2,4] sum=10 ≥ 7
 * Contract: [1,2,4] sum=7 ≥ 7
 * Contract: [2,4] sum=6 < 7
 * Expand: [2,4,3] sum=9 ≥ 7
 * Contract: [4,3] sum=7 ≥ 7, length=2
 *
 * Output:
 * ```
 * 2 (minimum length)
 * ```
 *
 * ### TIME COMPLEXITY:
 * O(n)**
 * - Single pass through input
 *
 * ### SPACE COMPLEXITY:
 * **O(n)** - [Explanation of why this complexity]. The algorithm [describe the operation] which takes **O(n)** space.
 *
 * ### EDGE CASES:
 * - **Empty input**: Handle when input is empty
 * - **Single element**: Handle single-element inputs
 * - **Boundary values**: Handle minimum/maximum valid values
 *
 * *
 */

/**
 * Main solution for Problem 209: Minimum Size Subarray Sum
 *
 * @param {number} target - Target sum to achieve or exceed
 * @param {number[]} nums - Array of positive integers
 * @return {number} - Minimum length of subarray with sum >= target, or 0 if none exists
 *
 * Time Complexity: O(n)
 * Space Complexity: O(1)
 */
function solve(target, nums) {
  if (!nums || nums.length === 0) return 0;

  let left = 0;
  let sum = 0;
  let minLength = Infinity;

  for (let right = 0; right < nums.length; right++) {
    // Expand window by adding current element
    sum += nums[right];

    // Contract window while sum is still >= target
    while (sum >= target) {
      minLength = Math.min(minLength, right - left + 1);
      sum -= nums[left];
      left++;
    }
  }

  return minLength === Infinity ? 0 : minLength;
}

/**
 * Test cases for Problem 209: Minimum Size Subarray Sum
 */
function testSolution() {
  console.log("Testing 209. Minimum Size Subarray Sum");

  // Test case 1: Basic example
  const result1 = solve(7, [2, 3, 1, 2, 4, 3]);
  const expected1 = 2;
  console.assert(
    result1 === expected1,
    `Test 1 failed: expected ${expected1}, got ${result1}`,
  );
  console.log(`Test 1 passed: target=7, nums=[2,3,1,2,4,3] -> ${result1}`);

  // Test case 2: Single element sufficient
  const result2 = solve(4, [1, 4, 4]);
  const expected2 = 1;
  console.assert(
    result2 === expected2,
    `Test 2 failed: expected ${expected2}, got ${result2}`,
  );
  console.log(`Test 2 passed: target=4, nums=[1,4,4] -> ${result2}`);

  // Test case 3: No valid subarray
  const result3 = solve(11, [1, 1, 1, 1, 1, 1, 1, 1]);
  const expected3 = 0;
  console.assert(
    result3 === expected3,
    `Test 3 failed: expected ${expected3}, got ${result3}`,
  );
  console.log(`Test 3 passed: target=11, nums=[1,1,1,1,1,1,1,1] -> ${result3}`);

  // Test case 4: Entire array needed
  const result4 = solve(15, [1, 2, 3, 4, 5]);
  const expected4 = 5;
  console.assert(
    result4 === expected4,
    `Test 4 failed: expected ${expected4}, got ${result4}`,
  );
  console.log(`Test 4 passed: target=15, nums=[1,2,3,4,5] -> ${result4}`);

  // Test case 5: Empty array
  const result5 = solve(5, []);
  const expected5 = 0;
  console.assert(
    result5 === expected5,
    `Test 5 failed: expected ${expected5}, got ${result5}`,
  );
  console.log(`Test 5 passed: target=5, nums=[] -> ${result5}`);

  // Test case 6: Single element array - sufficient
  const result6 = solve(3, [5]);
  const expected6 = 1;
  console.assert(
    result6 === expected6,
    `Test 6 failed: expected ${expected6}, got ${result6}`,
  );
  console.log(`Test 6 passed: target=3, nums=[5] -> ${result6}`);

  // Test case 7: Single element array - insufficient
  const result7 = solve(10, [5]);
  const expected7 = 0;
  console.assert(
    result7 === expected7,
    `Test 7 failed: expected ${expected7}, got ${result7}`,
  );
  console.log(`Test 7 passed: target=10, nums=[5] -> ${result7}`);

  // Test case 8: Large numbers
  const result8 = solve(213, [12, 28, 83, 4, 25, 26, 25, 2, 25, 25, 25, 12]);
  const expected8 = 8;
  console.assert(
    result8 === expected8,
    `Test 8 failed: expected ${expected8}, got ${result8}`,
  );
  console.log(
    `Test 8 passed: target=213, nums=[12,28,83,4,25,26,25,2,25,25,25,12] -> ${result8}`,
  );

  console.log("All test cases passed for 209. Minimum Size Subarray Sum!");
}

/**
 * Example usage and demonstration
 */
function demonstrateSolution() {
  console.log("\n=== Problem 209. Minimum Size Subarray Sum ===");
  console.log("Category: Sliding Window");
  console.log("Difficulty: Medium");
  console.log("");

  testSolution();
}

// Run tests if this file is executed directly
if (require.main === module) {
  demonstrateSolution();
}

// Export for use in other modules
module.exports = {
  solve,
  testSolution,
  demonstrateSolution,
};

/**
 * Additional Notes:
 * - This solution focuses on sliding window concepts
 * - Consider the trade-offs between time and space complexity
 * - Edge cases are crucial for robust solutions
 * - The approach can be adapted for similar problems in this category
 */

File: solutions/sliding-window/python/0209-minimum-size-subarray-sum.py
"""
# Difficulty: Medium

# 0209. Minimum Size Subarray Sum

Given an array of positive integers nums and a positive integer target, return the minimal length of a subarray whose sum is greater than or equal to target. If there is no such subarray, return 0 instead.

**Example:**

<dl class="example-details">
<dt>Input:</dt>
<dd>target = 7, nums = [2,3,1,2,4,3]</dd>
<dt>Output:</dt>
<dd>2 (minimum length)</dd>
<dt>Explanation:</dt>
<dd>The minimal length subarray with sum ≥ 7 is [4,3] with length 2</dd>
</dl>

<details>
<summary><b>🔍 SOLUTION EXPLANATION</b></summary>
### METADATA:
**Techniques**: Hash Table Lookup, Array Traversal, Sliding Window
**Data Structures**: Hash Map, Hash Set, Array
**Patterns**: Complement Search, Two Pointers Pattern
**Time Complexity**: O(n) - Single pass through input
**Space Complexity**: O(1) - Constant extra space

### INTUITION:
The key insight is that use sliding window. Expand window until sum >= target. Then shrink from left while sum >= target. Track minimum length. This achieves O(n) time with single pass.

### APPROACH:
1. **Initialize variables**: Set left = 0, min_len = float('inf'), current_sum = 0
2. **Expand with right pointer**: For right in range(len(nums))
3. **Add to window**: current_sum += nums[right]
4. **Contract while valid**: While current_sum >= target
5. **Update minimum**: min_len = min(min_len, right - left + 1)
6. **Shrink window**: current_sum -= nums[left], left += 1
7. **Continue scanning**: Process all elements
8. **Return result**: Return min_len if found, else 0

### WHY THIS WORKS:
- This ensures that sliding window expands right until sum >= target, then contracts left
- This ensures that greedy contraction: shrink window while maintaining sum >= target
- This ensures that each element added once (right++) and removed once (left++)
- This ensures that track minimum window size satisfying sum condition
- This ensures that o(n) time: two pointers scan array once, O(1) space

### EXAMPLE WALKTHROUGH:
Input:
```
target = 7, nums = [2,3,1,2,4,3]
```

Step 1: Expand window
[2,3,1,2] sum=8 ≥ 7
Step 2: Contract
[3,1,2] sum=6 < 7
Expand: [3,1,2,4] sum=10 ≥ 7
Contract: [1,2,4] sum=7 ≥ 7
Contract: [2,4] sum=6 < 7
Expand: [2,4,3] sum=9 ≥ 7
Contract: [4,3] sum=7 ≥ 7, length=2

Output:
```
2 (minimum length)
```

### TIME COMPLEXITY:
O(n)
- Single pass through input


### SPACE COMPLEXITY:
O(1)
- Constant extra space


### EDGE CASES:
- **Empty input**: Handle when input is empty
- **Single element**: Handle single-element inputs
- **Boundary values**: Handle minimum/maximum valid values

</details>
"""

from typing import List, Optional, Dict, Tuple
import sys


class Solution:
    def minSubArrayLen(self, target: int, nums: List[int]) -> int:
        """
        Find the minimum length of a contiguous subarray whose sum is greater than
        or equal to the target.

        Args:
            target: The target sum we're looking for
            nums: List of positive integers

        Returns:
            int: Minimum length of subarray that sums to >= target, or 0 if no such subarray exists
        """
        if not nums:
            return 0

        n = len(nums)
        min_length = sys.maxsize
        current_sum = 0
        left = 0

        # Sliding window approach
        for right in range(n):
            # Add the current number to our running sum
            current_sum += nums[right]

            # While our sum is >= target, try to minimize the window
            while current_sum >= target:
                # Update minimum length if current window is smaller
                min_length = min(min_length, right - left + 1)

                # Remove leftmost element and shrink window
                current_sum -= nums[left]
                left += 1

        # Return 0 if we never found a valid subarray
        return min_length if min_length != sys.maxsize else 0


def test_solution() -> None:
    """
    Test cases for the solution.
    """
    solution = Solution()

    # Test case 1: Example from problem
    result = solution.minSubArrayLen(7, [2, 3, 1, 2, 4, 3])
    expected = 2
    assert result == expected, f"Expected expected, got result"

    # Test case 2: Empty array
    result = solution.minSubArrayLen(7, [])
    expected = 0
    assert result == expected, f"Expected expected, got result"

    print("All test cases passed!")


if __name__ == "__main__":
    test_solution()

    # Example usage
    solution = Solution()
    print(f"Solution for 209. Minimum Size Subarray Sum")
